### PR TITLE
[FLOC-4321] A couple __eq__ optimization for PClass

### PR DESCRIPTION
**Pull request by sarum90**
Some optimizations for `__eq__` in PClass.

Also some additional tests to get full code coverage / ensure we are not regressing with this change.

### CODE REVIEW

File: pyrsistent/_pclass.py [hunk 1]
```diff
@@ -135,7 +135,14 @@ def transform(self, *transformations):
         return transform(self, transformations)
 
     def __eq__(self, other):
+        if self is other:
+            return True
         if isinstance(other, self.__class__):
+            try:
```
Comment (moshez) [inline pyrsistent/_pclass.py]: Can we break this down further?

```
def __eq__(self, other):
      if self is other:
           return True
     if not isinstance(other, self.__class__):
           return False
     try:
          self_hash, other_hash = hash(self), hash(other)
     except:
          pass
     else:
          if self_hash != other_hash:
               return False
      for name in .....
```

This way you only protect the minimum amount of lines with try/except:

Comment (sarum90) [inline pyrsistent/_pclass.py]: Done. +1 for reducing content inside of an `except:` block.

**Comment by sarum90** [comment]
Waiting for travis before I merge.

**Comment by sarum90** [comment]
Only travis failure is because I don't have coveralls set up for this repo.

Note that looking at the coverage report there are 0 more missed branches, and all other tests pass.

Merging presently.

**Comment by sarum90** [comment]
We might want to restore this branch and push it upstream sometime.

**Pull request merged**